storage/src/adb/any/sync/client.rs

```diff
@@ -0,0 +1,638 @@
+                                "InvalidState: next_op_pos < lower_bound_ops"
+                            );
+                            Error::InvalidState
+                        })?;
+
+                debug!(
+                    lower_bound_ops = config.lower_bound_ops,
+                    upper_bound_ops = config.upper_bound_ops,
+                    total_ops_needed = total_ops_needed,
+                    next_op_pos = next_op_loc,
+                    applied_ops = applied_ops,
+                    "Calculating remaining operations"
+                );
+
+                let remaining_ops = total_ops_needed.checked_sub(applied_ops).ok_or_else(|| {
+                    warn!(
+                        total_ops_needed,
+                        applied_ops, "InvalidState: applied_ops > total_ops_needed"
+                    );
+                    Error::InvalidState
+                })?;
+
+                let batch_size = std::cmp::min(config.max_ops_per_batch.get(), remaining_ops);
+                let batch_size = NonZeroU64::new(batch_size).ok_or(Error::InvalidState)?;
+
+                debug!(
+                    target_hash = ?config.target_hash,
+                    lower_bound_pos = config.lower_bound_ops,
+                    upper_bound_pos = config.upper_bound_ops,
+                    next_op_loc = next_op_loc,
+                    remaining_ops = remaining_ops,
+                    batch_size = batch_size.get(),
+                    "Fetching proof and operations"
+                );
+
+                // Get proof and operations from resolver.
+                let (proof, new_operations) =
+                    config.resolver.get_proof(next_op_loc, batch_size).await?;
+                let new_operations_len = new_operations.len() as u64;
+
+                // Validate that we didn't get more operations than requested
+                // or that we didn't get an empty proof. We should never get an empty proof
+                // because we will never request an empty proof (i.e. a proof over an empty database).
+                if new_operations_len > batch_size.get() || new_operations_len == 0 {
+                    metrics.invalid_batches_received += 1;
+                    if metrics.invalid_batches_received > config.max_retries {
+                        return Err(Error::MaxRetriesExceeded);
+                    }
+                    return Ok(Client::FetchData {
+                        config,
+                        log,
+                        pinned_nodes,
+                        metrics,
+                    });
+                }
+
+                debug!(
+                    num_ops = new_operations_len,
+                    "Received operations from resolver"
+                );
+
+                // Verify the proof is valid over the given operations
+                if !adb::any::Any::<E, K, V, H, T>::verify_proof(
+                    &mut config.hasher,
+                    &proof,
```

Comment:
We should add a function on the resolver to indicate whether a proof result was successful or not. We can use this to block malicious peers.